routers/repo/view.go

```diff
@@ -36,6 +36,91 @@ const (
 	tplMigrating base.TplName = "repo/migrating"
 )
 
+type namedBlob struct {
+	name string
+	blob *git.Blob
+}
+
+func followLinks(entry *git.TreeEntry) (*git.TreeEntry, error) {
+	var err error
+	for i := 0; i < 999; i++ {
+		if entry.IsLink() {
+			entry, err = entry.FollowLink()
+			if err != nil {
+				return nil, err
+			}
+		} else {
+			break
+		}
+	}
+	if entry.IsRegular() || entry.IsExecutable() {
+		return entry, nil
+	}
+	return nil, nil
+}
+
+// FIXME: There has to be a more efficient way of doing this
+func getReadmeFileFromPath(commit *git.Commit, treePath string) (*namedBlob, error) {
+	tree, err := commit.SubTree(treePath)
+	if err != nil {
+		return nil, err
+	}
+
+	entries, err := tree.ListEntries()
+	if err != nil {
+		return nil, err
+	}
+
+	var readmeFiles [4]*namedBlob
+	var exts = []string{".md", ".txt", ""} // sorted by priority
+	for _, entry := range entries {
+		if entry.IsDir() {
+			continue
+		}
+		for i, ext := range exts {
+			if markup.IsReadmeFile(entry.Name(), ext) {
+				if readmeFiles[i] == nil || base.NaturalSortLess(readmeFiles[i].name, entry.Blob().Name()) {
+					name := entry.Name()
+					entry, err := followLinks(entry)
+					if err != nil {
+						return nil, err
+					}
+					if entry != nil {
+						readmeFiles[i] = &namedBlob{
+							name,
+							entry.Blob(),
+						}
+					}
+				}
+			}
+		}
+
+		if markup.IsReadmeFile(entry.Name()) {
+			if readmeFiles[3] == nil || base.NaturalSortLess(readmeFiles[3].name, entry.Blob().Name()) {
+				name := entry.Name()
+				entry, err := followLinks(entry)
```

Comment:
I would have been happy just skipping them. 😅 

Now I can produce the following:

```
$ ls -l
total 4
-rw-rw-r-- 1 gittest gittest 45 Feb 18 19:42 NOREAD.md
lrwxrwxrwx 1 gittest gittest 55 Feb 21 00:05 README.md -> ../../../../../../../../etc/passwd
```

```
500 Server Error
README.md: points outside of repo
```

Perhaps we should just (log and) ignore entries that produce any errors while attempting to render the README.

On the other hand, this is correctly handled (i.e. not rendered):
![image](https://user-images.githubusercontent.com/18600385/75001083-29d94880-543f-11ea-8146-761a287c1b8c.png)

Reply:
I've handled ErrBadLink better now.